충남대 FE 정지원 - 3단계 단체 주문 기능 구현하기

jjw5655 opened the pull request:
받는사람 추가버튼을 누르고 여러명을 선택할 수 있도록 구현했는데, 입력칸을 클릭하면 클릭대기하는 | 모양이 생겼다가 바로 없어지는 문제가 생겼습니다. 아래 이미지가 모달창인데, 빨간박스부분이 무한렌더링 되고있는 것 같아요. 
<img width="885" height="468" alt="image" src="https://github.com/user-attachments/assets/3b417eff-427f-4505-9349-1a899a33044d" />

해결해보려 했지만 잘 안되서 마지막 commit(fix: 받는사람 모달 무한 렌더링 문제 해결)에서 간단한 형태로 수정해서 일단 제출했습니다. 기존 코드에 어디가 문제인걸까요?

jjw5655 committed:
feat: React Hook Form 설치 및 주문 타입 추가

- package.json에 react-hook-form 의존성 추가 (npm install로 설치)
- types/order.ts에 주문 관련 타입 정의 (Recipient, OrderFormData 등)
- 전화번호 유효성 검사를 01012341234 형식만 허용하도록 변경
- 이름, 수량 유효성 검사 유틸리티 함수 추가
- 받는사람 중복 검사 및 유효성 검증을 위한 유틸리티 함수 생성
- index 파일에서 새로운 타입 및 유틸리티 export

다중 받는사람 기능을 React Hook Form으로 구현하기 위한 기반 구조 마련

jjw5655 committed:
refactor: 주문 폼을 React Hook Form으로 리팩토링

- useOrderForm hook을 useState에서 useForm으로 완전 전환
- OrderPage의 모든 입력 필드를 register()와 연결
- React Hook Form validation 시스템으로 유효성 검사 이전
- 불필요한 setter 핸들러 함수들 제거
- 기존 단일 받는사람 기능 완전 호환 유지

더 적은 코드로 동일한 기능을 구현하고 성능 개선

jjw5655 committed:
feat: 다중 받는사람 관리를 위한 기반 구조 구축

- useMultipleRecipientsForm hook 추가 (useFieldArray 활용)
- useRecipientModal hook으로 모달 상태 관리
- modalUtils, formUtils에 재사용 가능한 유틸리티 함수 추가
- 받는사람 관련 추가 타입 정의 (RecipientActions, OrderSummary 등)
- 최대 10명 제한, 전화번호 중복 검사, 총 수량 계산 기능
- ESC 키, 외부 클릭, 스크롤 방지 등 모달 UX 개선

jjw5655 committed:
feat: 받는사람 표시 컴포넌트 구현

- RecipientCard 컴포넌트로 개별 받는사람 정보 카드 UI 구현
- RecipientList 컴포넌트로 받는사람 목록 관리 기능 구현
- 받는사람 추가/삭제 버튼 및 최대 10명 제한 UI
- 빈 상태 처리 및 총 수량 요약 정보 표시
- OrderPage에서 기존 단일 폼을 다중 받는사람 목록으로 교체
- 카카오 스타일 디자인 및 호버 효과 적용

jjw5655 committed:
feat: 받는사람 추가 모달 컴포넌트 구현

- 재사용 가능한 Modal 컴포넌트 (ESC/외부클릭/스크롤방지)
- RecipientForm으로 개별 받는사람 정보 입력 폼 구현
- RecipientModal로 다중 받는사람 관리 모달 구현
- "추가하기" 버튼으로 받는사람 폼 동적 추가 (최대 10명)
- 실시간 유효성 검사 및 전화번호 중복 검사
- OrderPage에서 RecipientList와 모달 연동
- "n명 완료" 버튼으로 받는사람 일괄 저장 기능

jjw5655 committed:
feat: 다중 받는사람 관리 시스템 통합 완성

- useOrderForm을 useMultipleRecipientsForm으로 완전 교체
- 받는사람 목록 상태 관리를 중앙 집중화
- 총 수량 및 가격 계산 로직 통합 (받는사람별 수량 합계)
- 주문 버튼에 총 가격 표시 및 비활성화 상태 추가
- 완전한 폼 유효성 검사 및 에러 처리 구현
- 최대 10명 제한, 중복 전화번호 검사 등 모든 요구사항 완성

jjw5655 committed:
feat: 전화번호 유효성 검사 개선 및 중복 검사 강화

- phoneUtils.ts 유틸리티 파일 추가
  - 전화번호 정규화 함수 (normalizePhoneNumber)
  - 엄격한 전화번호 검증 함수 (isValidPhoneNumberStrict)
- 전화번호 중복 검사 로직 개선
  - 정규화된 전화번호로 더 정확한 중복 검사
  - 포맷에 관계없이 일관된 중복 검사 수행
- 전화번호 유효성 검사 메시지 개선
  - 더 구체적인 에러 메시지 제공
  - 11자리 길이 검증 강화

jjw5655 committed:
fix: 받는사람 모달 무한 렌더링 문제 해결

- RecipientModal을 단순한 단일 입력 형태로 변경
- existingRecipients prop 제거로 무한 렌더링 원인 해결
- 복잡한 useEffect, validation 로직 제거
- useState 3개만 사용하는 최소한의 구조로 단순화

greenblues1190 commented on src/components/order/RecipientModal.tsx:
```diff
@@ -0,0 +1,227 @@
+  `}
+`;
+
+const RecipientModal = ({ isOpen, onClose, onSave }: RecipientModalProps) => {
```
`initialRecipients`의 기본값으로 할당한 빈배열 리터럴(`[]`)이 컴포넌트 리렌더링 시마다 새로 생성되면서 useEffect를 다시 트리거 -> 상태 변경 -> 새로 생성 -> useEffect 트리거 -> 무한 반복되었네요

```tsx
const RecipientModal = ({
  initialRecipients = [],
  ...
}: RecipientModalProps) => {
```

컴포넌트 외부에서 상수로 선언하여 재생성되는 일이 없도록 해야할 듯 합니다

```tsx
const EMPTY_RECIPIENTS: Recipient[] = [];

const RecipientModal = ({
  initialRecipients = EMPTY_RECIPIENTS,
  ...
}: RecipientModalProps) => {
```

greenblues1190 commented on src/components/common/Modal.tsx:
```diff
@@ -0,0 +1,174 @@
+    return () => {
+      document.body.style.overflow = 'unset';
+    };
+  }, [isOpen]);
```
- 커스텀훅으로 분리하면 재사용하기 좋은 유틸 훅으로 보여요 💯
- 모달을 열기 전 body의 overflow 속성이 unset이 아닐 수도 있다면 어떻게 될까요?

jjw5655 replied:
스크롤 방지 훅으로 분리했습니다. 지금 코드는 모달을 닫을 때 항상 'unset'으로 변경하므로 문제가 생길 수 있어서 이전 값을 저장했다가 닫을 때 복구시키는 방식으로 코드를 수정했습니다.

greenblues1190 commented on src/pages/OrderPage.tsx:
```diff
@@ -120,49 +123,69 @@ const OrderButtonBar = styled.div`
+  &:hover:not([disabled]) {
+    background: ${(props) => (props.disabled ? '#f3f4f6' : '#fde047')};
+  }
 `;
```
div 보다는 button요소로 생성하는 것이 좋겠네요

greenblues1190 commented on src/pages/OrderPage.tsx:
```diff
@@ -120,49 +123,69 @@ const OrderButtonBar = styled.div`
+    !messageError &&
+    !senderError &&
+    !recipientsError &&
+    recipients.every((r) => r.name.trim() && r.phone.trim() && r.quantity >= 1);
```
formState.isValid를 사용하지 않은 이유가 있으신가요?

jjw5655 replied:
리액트훅폼을 사용하도록 수정했습니다!

greenblues1190 commented on src/hooks/useRecipientModal.ts:
사용되고 있는 훅인가요?

jjw5655 replied:
사용되지 않네요. 삭제했습니다.

greenblues1190 commented on src/hooks/useOrderForm.ts:
사용되고 있는 훅인가요?

jjw5655 replied:
마찬가지로 삭제했습니다.

greenblues1190 commented on src/components/order/RecipientForm.tsx:
```diff
@@ -0,0 +1,246 @@
+  existingRecipients = [],
+  disabled = false,
+}: RecipientFormProps) => {
+  const [formData, setFormData] = useState(
+    () => initialData || createEmptyRecipientForm()
+  );
+  const [errors, setErrors] = useState<{ [key: string]: string }>({});
+
+  // 폼 데이터 유효성 검사 및 상위 컴포넌트로 전달
+  useEffect(() => {
+    const validation = validateRecipientForm(formData);
+    let newErrors = { ...validation.errors };
+
+    // 전화번호 중복 검사 (기존 받는사람 목록과 비교)
+    const normalizedCurrentPhone = normalizePhoneNumber(formData.phone);
+    if (
+      normalizedCurrentPhone &&
+      existingRecipients.some(
+        (r) => normalizePhoneNumber(r.phone) === normalizedCurrentPhone
+      )
+    ) {
+      newErrors.phone = '이미 등록된 전화번호입니다.';
+    }
+
+    setErrors(newErrors);
+
+    // 저장할 때는 정규화된 전화번호로 저장
+    const dataToSave = {
+      ...formData,
+      phone: normalizedCurrentPhone,
+    };
+    onDataChange(index, dataToSave);
+  }, [formData, index, onDataChange, existingRecipients]);
```
지금은 OrderPage -> RecipientModal -> RecipientForm으로 폼 제어 관련 props를 깊게 내려주고 있어요.
모달 내 받는 사람 등록은 사실 주문 페이지의 폼과 독립된 별도 폼이니 아래 구조로 분리해보면 어떨까요?

- OrderPage: 축하 메시지,보내는 사람 이름, receivers 배열만 관리하는 useForm
- RecipientModal(RecipientForm): receivers 배열 전용 별도 useForm
  - 모달에서 `n명 완료` submit 버튼 클릭 시 모달 폼의 handleSubmit으로 receivers 배열을 반환
  - OrderPage는 setValue("receivers", returnedArray)로 반환받은 받는 사람 정보를 반영

jjw5655 replied:
말씀하신 대로 수정했습니다!

greenblues1190 commented on src/components/order/RecipientForm.tsx:
```diff
@@ -0,0 +1,246 @@
+    }));
+  };
+
+  const handleRemove = () => {
+    onRemove(index);
+  };
```
[useFieldArray](https://react-hook-form.com/docs/usefieldarray) 훅으로 다시 수정해주세요!

greenblues1190 reviewed (changes requested):
안녕하세요 지원님 리뷰가 늦었습니다 🙇

커밋 기록을 보니 먼저 기반 구조를 완성한 뒤 UI를 덧붙이는 순서로 작업하셨네요.
실무에서는 UI에 기능이 동작하도록 빠르게 구현한 뒤, 그 다음에 추상화, 리팩터링을 진행하는 방식을 추천드립니다.
이렇게 하면 초반 설계 예측이 빗나가더라도, 후반에 발생할 수 있는 사이드이펙트를 최소화할 수 있어요.
수고하셨습니다 말씀주신 버그는 아래 코멘트로 피드백드렸어요

jjw5655 committed:
Revert "fix: 받는사람 모달 무한 렌더링 문제 해결"

This reverts commit a40e79d06fa5f5641ec320c67f3fb8810430b50d.

jjw5655 committed:
fix: 받는사람 모달 무한 렌더링 문제 해결

- RecipientModal props의 기본값인 빈 배열 리터럴([])이 리렌더링 시마다 재생성되어 useEffect의 무한 호출을 유발하는 문제 있었음.
- 컴포넌트 외부에 상수를 선언하여 동일한 참조를 유지하도록 수정, 불필요한 렌더링 방지함.

jjw5655 committed:
feat: 받는 사람 폼 유효성 검사 로직 개선

- 모달이 열리자마자 유효성 검사가 실행되어, 사용자 입력 전에 오류가 표시되는 사용자 경험 문제 개선.
- 'touched' 상태를 추가하여 필드 상호작용 여부를 추적함.
- 사용자가 입력 필드를 벗어났을 때(onBlur)만 유효성 검사를 실행하도록 로직 변경.

jjw5655 committed:
feat: body 스크롤 제어 로직을 재사용 가능한 커스텀 훅으로 분리

- Modal 컴포넌트의 body 스크롤 제어 로직을 useBodyScrollLock 커스텀 훅으로 분리
- 기존 body overflow 값을 저장 후 복원하는 안전한 방식으로 구현하여 예기치 않은 스타일 변경 방지
- 13줄의 useEffect 로직을 1줄의 훅 호출로 간소화

jjw5655 committed:
refactor: 주문 버튼을 div에서 button 요소로 변경

jjw5655 committed:
refactor: React Hook Form의 formState.isValid 활용하여 주문 가능 여부 판단 로직 개선

- 수동으로 작성된 복잡한 유효성 검사 조건을 formState.isValid로 대체
- useMultipleRecipientsForm 훅에서 formState 반환하도록 수정
- OrderPage에서 formState.isValid를 활용하여 코드 간소화

jjw5655 committed:
chore: 사용되지 않는 커스텀 훅 제거

- useRecipientModal, useOrderForm 훅 파일 삭제
- 두 훅 모두 어떤 컴포넌트에서도 사용되지 않는 데드 코드였음
- hooks 배럴파일(index.ts)에서 해당 export 제거

jjw5655 committed:
refactor: 폼 상태 관리를 관심사별로 분리

- OrderPage: 주문 정보(message, sender, recipients)만 관리하는 useForm으로 단순화
- RecipientModal: 받는사람 전용 useForm + useFieldArray로 독립적 관리
- RecipientForm: React Hook Form의 useController로 개별 필드 제어
- 데이터 흐름: Modal의 handleSubmit → OrderPage의 setValue로 깔끔한 연동
- useMultipleRecipientsForm 복잡한 훅 제거로 코드 간소화

Props drilling 해결 및 각 컴포넌트의 책임 명확화로 유지보수성 크게 향상

jjw5655 committed:
docs: README를 3단계 단체 주문 기능 구현 요구사항으로 업데이트

greenblues1190 reviewed (approved):
수고많으셨습니다 💯
